### PR TITLE
network: reject addresses where there is no kernel support.

### PR DESCRIPTION
**Pull request by htuch**
This support environments in which IPv4/IPv6 is not enabled at kernel
level.

Risk Level: Low
Testing: Manual; ran Envoy in our IPv6-only environment with IPv4
  config. Testing for this in CI is not simple, since we don't have
  an easy way to mock out kernel support.

Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

File: source/common/network/address_impl.cc [hunk 1]
```diff
@@ -19,6 +19,30 @@ namespace Envoy {
 namespace Network {
 namespace Address {
 
+namespace {
+
+// Validate that IPv4 is supported on this platform, raise an exception for the
+// given address if not.
+void validateIpv4Supported(const std::string& address) {
+  static bool supported = [] { return ::socket(AF_INET, SOCK_STREAM, 0) != -1; }();
```
Comment (ggreenway) [inline source/common/network/address_impl.cc]: This leaks a socket. Please close() if it's >= 0.

Comment (ggreenway) [inline source/common/network/address_impl.cc]: const?

**Comment by ggreenway** [comment]
Also, if you add ::socket() to OsSysCalls, you can mock it and add test coverage.

**Comment by htuch** [comment]
@ggreenway I like the idea of mocking via OsSysCalls, but the static/singleton combination doesn't work well with this pattern; the unmocked IP family support values are cached before we hit this test in a multi-test binary. We could just unconditionally check for IP family support on each address instance creation..

**Comment by ggreenway** [comment]
No, checking before each instance creation is both expensive (2 syscalls) and runs into the same problem as the original PR: it'll throw if the system runs out of sockets. I'm ok with not mocking ::socket() since it doesn't work well in this case.

**Comment by htuch** [comment]
Yes, I agree. Can you approve?

**Comment by ggreenway** [comment]
Given that mocking ::socket() for this test doesn't work (and can't work), I think you should revert the last 2 commits. Leaving it as using OsSysCalls means someone else might try to mock that and break multi-test binaries.

**Comment by mattklein123** [comment]
Couldn't we still test as long as we make a dedicated test binary for just the required test?

**Comment by htuch** [comment]
You would need to exclude it from coverage due to the unified binary. Let
me know what you prefer here, I don't mind either way, I would be more
strongly motivated if this was an actual regression that we wanted to catch
in future.

**Comment by alyssawilk** [comment]
Bleh.  As it turns out, over 100 Envoy tests don't properly support v6-only machines.  Things like the MockHostDescription do resolveUrl(harcodedv4url) which now breaks, though we never create a socket around it.

I can fix up the tests this week but I think this change is going to be pretty hard on the v6-only user[s].   CircleCi currently protects us from us breaking things for v4-only users (by not supporting v6 by default)  but I suspect there will be regressions of people using v4-only addresses in other places which will only be caught when Google merges this code downstream.  If anyone has ideas of how to catch more of these in review/CI it'd be much appreciated!

**Comment by htuch** [comment]
I'm OK with a rollback of this PR as it's not critical (it was intended to unblock some internal tests in the first place) and I'm busyish this week at a summit.

I think we could roll forward with a new CI job that runs with the use of Api::OsSyscalls for `socket(2)` that allows us to simulate a pure IPv6 environment. Of course, if we can get CircleCI to provide us a pure IPv6 environment which will fail `socket(2)` for IPv4, that would be even better.

**Comment by alyssawilk** [comment]
Short term I can just remove the throw() on our import, so it's not a huge problem.  But I'd like to remove it eventually so I'd just like a long term where we don't keep fixing up broken ipv4 tests :-)

**Pull request merged**